src/test/java/com/yelp/nrtsearch/server/backup/BackupDiffManagerTest.java

```diff
@@ -0,0 +1,389 @@
+import java.util.stream.Collectors;
+import org.junit.*;
+import org.junit.rules.TemporaryFolder;
+
+public class BackupDiffManagerTest {
+
+  private BackupHelper backupHelper;
+
+  public static class BackupDiffInfoTest {
+    @Test
+    public void getBackupDiffInfo() {
+      Set<String> oldFileNames = Set.of("1", "2", "3", "4");
+      Set<String> currentFileNames = Set.of("1", "4", "5");
+      BackupDiffManager.BackupDiffInfo backupDiffInfo =
+          BackupDiffManager.BackupDiffInfo.generateBackupDiffInfo(oldFileNames, currentFileNames);
+      assertEquals(Set.of("5"), backupDiffInfo.getToBeAdded());
+      assertEquals(Set.of("1", "4"), backupDiffInfo.getAlreadyUploaded());
+      assertEquals(Set.of("2", "3"), backupDiffInfo.getToBeRemoved());
+    }
+  }
+
+  public static class TempDirManagerTest {
+    @Rule public final TemporaryFolder folder = new TemporaryFolder();
+    private Path someTempDir;
+
+    @Before
+    public void setUp() throws IOException {
+      someTempDir = folder.newFolder("some_temp_dir").toPath();
+    }
+
+    @Test
+    public void tempDirCreateAndDelete() throws Exception {
+      Path tempDir;
+      try (BackupDiffManager.TempDirManager tempDirManager =
+          new BackupDiffManager.TempDirManager(someTempDir)) {
+
+        Path tempDirPath = Paths.get(tempDirManager.getPath().toString());
+        Path someTempFile = Paths.get(tempDirPath.toString(), "someTempFile");
+        assertEquals(false, Files.exists(tempDirPath));
+        Files.createDirectory(tempDirPath);
+        Files.createFile(someTempFile);
+        assertEquals(true, Files.exists(tempDirPath));
+        assertEquals(true, Files.exists(someTempFile));
+        tempDir = tempDirManager.getPath();
+      }
+      assertEquals(false, Files.exists(tempDir));
+    }
+
+    @Test
+    public void tempDirDelete() throws Exception {
+      Path tempDir;
+      try (BackupDiffManager.TempDirManager tempDirManager =
+          new BackupDiffManager.TempDirManager(someTempDir)) {
+
+        Path tempDirPath = Paths.get(tempDirManager.getPath().toString());
+        assertEquals(false, Files.exists(tempDirPath));
+        tempDir = tempDirManager.getPath();
+      }
+      assertEquals(false, Files.exists(tempDir));
+    }
+  }
+
+  public static class BackupDiffMarshallerTest {
+    @Rule public final TemporaryFolder folder = new TemporaryFolder();
+
+    @Test
+    public void testSerDe() throws IOException {
+      List<String> indexFileNames = List.of("1", "2", "3");
+      final Path tempFile = folder.newFile("tempFile.txt").toPath();
+      BackupDiffManager.BackupDiffMarshaller.serializeFileNames(indexFileNames, tempFile);
+      assertEquals(
+          indexFileNames, BackupDiffManager.BackupDiffMarshaller.deserializeFileNames(tempFile));
+    }
+  }
+
+  public static class BackupDiffDirValidatorTest {
+    @Rule public final TemporaryFolder folder = new TemporaryFolder();
+
+    @Test(expected = RuntimeException.class)
```

Comment:
Are you suggesting using a `CheckedException` here instead? If so, one issue with that approach is if the `validateMetaFile` method that it tests fails it is a hard failure, that is we cannot download any index data. Thus the caller would need to eventually throw a RuntimeException anyway. Let me know if you are thinking of something else.

Reply:
No, I meant using a class that extends `RuntimeException` rather than `RuntimeException` itself to be more specific. If not create a new class can at least use exceptions like `IllegalStateException`, `IllegalArgumentException`, etc. With RuntimeException there is a higher possibility that tests still pass if an exception is thrown from somewhere other than the intended place.

Reply:
fixed